### PR TITLE
partially ported pt methods and classes of vit mae to tensorflow.

### PR DESCRIPTION
**Pull request by sayakpaul**
~Porting is likely done till `TFViTMAEOutput`.~

Porting complete.

### CODE REVIEW

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py [hunk 1]
```diff
@@ -0,0 +1,961 @@
...
+    Args:
+        embed_dim (`int`):
+            Embedding dimension.
+        grid_size (`int`):
+            The grid height and width.
+        add_cls_token (`bool`, *optional*, defaults to `False`):
+            Whether or not to add a classification (CLS) token.
+
+    Returns:
+        (`torch.FloatTensor` of shape (grid_size*grid_size, embed_dim) or (1+grid_size*grid_size, embed_dim): the
```
Comment (ariG23498) [inline src/transformers/models/vit_mae/modeling_tf_vit_mae.py:70]:
```suggestion
        (`tf.Tensor` of shape (grid_size*grid_size, embed_dim) or (1+grid_size*grid_size, embed_dim): the
```

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py [hunk 2]
```diff
@@ -0,0 +1,961 @@
...
+        position embeddings (with or without classification token)
+    """
+    grid_h = tf.range(grid_size, dtype=tf.float32)
+    grid_w = tf.range(grid_size, dtype=np.float32)
```
Comment (ariG23498) [inline src/transformers/models/vit_mae/modeling_tf_vit_mae.py:74]:
```suggestion
    grid_w = tf.range(grid_size, dtype=tf.float32)
```

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py [hunk 3]
```diff
@@ -0,0 +1,961 @@
...
+    return pos_embed
+
+
+def get_2d_sincos_pos_embed_from_grid(embed_dim, grid):
+    if embed_dim % 2 != 0:
+        raise ValueError("embed_dim must be even")
+
+    # use half of dimensions to encode grid_h
+    emb_h = get_1d_sincos_pos_embed_from_grid(embed_dim // 2, grid[0])  # (H*W, D/2)
+    emb_w = get_1d_sincos_pos_embed_from_grid(embed_dim // 2, grid[1])  # (H*W, D/2)
```
Comment (ariG23498) [inline src/transformers/models/vit_mae/modeling_tf_vit_mae.py:91]:
```suggestion
    emb_h = get_1d_sincos_pos_embed_from_grid(embed_dim // 2, grid[0])  # (W*H, D/2)
    emb_w = get_1d_sincos_pos_embed_from_grid(embed_dim // 2, grid[1])  # (W*H, D/2)
```

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py [hunk 4]
```diff
@@ -0,0 +1,961 @@
...
+
+    emb = tf.concat([emb_h, emb_w], axis=1)  # (H*W, D)
```
Comment (ariG23498) [inline src/transformers/models/vit_mae/modeling_tf_vit_mae.py:93]:
```suggestion
    emb = tf.concat([emb_h, emb_w], axis=1)  # (W*H, D)
```

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py [hunk 5]
```diff
@@ -0,0 +1,961 @@
...
+    return emb
+
+
+def get_1d_sincos_pos_embed_from_grid(embed_dim, pos):
+    """
+    embed_dim: output dimension for each position pos: a list of positions to be encoded: size (M,) out: (M, D)
```
Comment (ariG23498) [inline src/transformers/models/vit_mae/modeling_tf_vit_mae.py:99]:
```suggestion
    embed_dim: output dimension for each position
    pos: a list of positions to be encoded: size (M,) out: (M, D)
```

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py [hunk 6]
```diff
@@ -0,0 +1,961 @@
...
+            shape=(1, num_patches + 1, self.config.hidden_size),
+            initializer="zeros",
+            trainable=False,  # fixed sin-cos embedding
+            name="position_embeddings",
+        )
+        pos_embed = get_2d_sincos_pos_embed(
+            self.position_embeddings.shape[-1],
+            int(self.patch_embeddings.num_patches ** 0.5),
+            add_cls_token=True,
+        )[None, ...]
```
Comment (ariG23498) [inline src/transformers/models/vit_mae/modeling_tf_vit_mae.py:150]:
```suggestion
        )[tf.newaxis, ...]
```

IMO this is more readable than None. WDYT?

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py [hunk 7]
```diff
@@ -0,0 +1,961 @@
...
+
+        super().build(input_shape)
+
+    def random_masking(self, sequence):
+        """
+        Perform per-sample random masking by per-sample shuffling. Per-sample shuffling is done by argsort random
+        noise.
+
+        Args:
+            sequence (`torch.LongTensor` of shape `(batch_size, sequence_length, dim)`)
```
Comment (ariG23498) [inline src/transformers/models/vit_mae/modeling_tf_vit_mae.py:161]:
```suggestion
            sequence (`tf.Tensor` of shape `(batch_size, sequence_length, dim)`)
```

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py [hunk 8]
```diff
@@ -0,0 +1,961 @@
...
+        # sort noise for each sample
+        ids_shuffle = tf.argsort(noise, axis=1)  # ascend: small is keep, large is remove
+        ids_restore = tf.argsort(ids_shuffle, axis=1)
+
+        # keep the first subset
+        ids_keep = ids_shuffle[:, :len_keep]
+        sequence_masked = tf.gather(
+            sequence,
+            axis=1,
+            indices=tf.repeat(ids_keep[..., None], [1, 1, dim]),
```
Comment (ariG23498) [inline src/transformers/models/vit_mae/modeling_tf_vit_mae.py:177]:
```suggestion
            indices=tf.repeat(ids_keep[..., tf.newaxis], [1, 1, dim]),
```

**Review by ariG23498 — CHANGES_REQUESTED** [review]
Kudos for the positional embedding code. Loved it!

Minor changes that need to be done in order for us to move ahead. I think we are heading in the right direction.

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py (hunk 3 above)
Comment (sayakpaul) [reply src/transformers/models/vit_mae/modeling_tf_vit_mae.py:91]: Not sure about this.

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py (hunk 4 above)
Comment (sayakpaul) [reply src/transformers/models/vit_mae/modeling_tf_vit_mae.py:93]: Not sure about this.

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py (hunk 6 above)
Comment (sayakpaul) [reply src/transformers/models/vit_mae/modeling_tf_vit_mae.py:150]: Conflicted between `tf.expand_dims()` then.

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py (hunk 8 above)
Comment (sayakpaul) [reply src/transformers/models/vit_mae/modeling_tf_vit_mae.py:177]: Same.

**Comment by sayakpaul** [comment]
@ariG23498 made the changes.

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py (hunk 3 above)
Comment (ariG23498) [reply src/transformers/models/vit_mae/modeling_tf_vit_mae.py:91]: If we take this [code line](https://github.com/ariG23498/transformers/blob/8d27eb36c120b032db80011c6179d462a38d30c5/src/transformers/models/vit_mae/modeling_tf_vit_mae.py#L75) under consideration, the `emb_h` and `emb_w` are width first.

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py (hunk 4 above)
Comment (ariG23498) [reply src/transformers/models/vit_mae/modeling_tf_vit_mae.py:93]: Same as above!

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py (hunk 6 above)
Comment (ariG23498) [reply src/transformers/models/vit_mae/modeling_tf_vit_mae.py:150]: `[tf.newaxis, ...]` is the closest to `[None, ...]`

If we want to have a more spelt out API for this change I would always like `tf.expand_dims()`

Comment (sayakpaul) [reply src/transformers/models/vit_mae/modeling_tf_vit_mae.py:150]: That in retrospect would further complicate the readability. One can add a separate line to have the dimensions expanded but that's unnecessary here because that's not gonna be reused.

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py (hunk 3 above)
Comment (sayakpaul) [reply src/transformers/models/vit_mae/modeling_tf_vit_mae.py:91]: https://github.com/ariG23498/transformers/blob/8d27eb36c120b032db80011c6179d462a38d30c5/src/transformers/models/vit_mae/modeling_tf_vit_mae.py#L78

**Review by ariG23498 — CHANGES_REQUESTED** [review]

**Comment by sayakpaul** [comment]
Some of the redundant stuff has been kept there on purpose. These mainly have two flavors:

* In large codebases like this, it helps to have uniform modules that have already been tested and therefore is more likely to not break. This often includes simple things too. 
* Some of the tensor shapes have been unpacked but there were never used. I think they are there for reference. Technically, this could have been added as a comment too. But I am just trying to convey some of the hard compromises you may have to make in order to impose maintainability, sanity, etc.

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py [hunk 9]
```diff
@@ -0,0 +1,1145 @@
...
+
+    def random_masking(self, sequence):
+        """
+        Perform per-sample random masking by per-sample shuffling. Per-sample shuffling is done by argsort random
+        noise.
+
+        Args:
+            sequence (`tf.Tensor` of shape `(batch_size, sequence_length, dim)`)
+        """
+        batch_size, seq_length, dim = sequence.shape
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/transformers/models/vit_mae/modeling_tf_vit_mae.py:252]: Advantage?

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py [hunk 10]
```diff
@@ -0,0 +1,1145 @@
...
+        # When running on CPU, `tf.keras.layers.Conv2D` doesn't support `NCHW` format.
+        # So change the input format from `NCHW` to `NHWC`.
+        # shape = (batch_size, in_height, in_width, in_channels=num_channels)
+        pixel_values = tf.transpose(pixel_values, perm=(0, 2, 3, 1))
+
+        projection = self.projection(pixel_values)
+
+        # Change the 2D spatial dimensions to a single temporal dimension.
+        # shape = (batch_size, num_patches, out_channels=embed_dim)
+        num_patches = (width // self.patch_size[1]) * (height // self.patch_size[0])
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/transformers/models/vit_mae/modeling_tf_vit_mae.py:348]: Taken from TFViT.

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py [hunk 11]
```diff
@@ -0,0 +1,1145 @@
...
+            units=config.hidden_size,
+            kernel_initializer=get_initializer(config.initializer_range),
+            name="dense",
+        )
+        self.dropout = tf.keras.layers.Dropout(rate=config.hidden_dropout_prob)
+
+    def call(
+        self,
+        hidden_states: tf.Tensor,
+        input_tensor: tf.Tensor,
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/transformers/models/vit_mae/modeling_tf_vit_mae.py:462]: Yeah, taken from TFViT.

**Comment by sayakpaul** [comment]
@ariG23498 made the further changes.

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ariG23498) [reply src/transformers/models/vit_mae/modeling_tf_vit_mae.py:348]: interesting -- do you think we should raise this concern in the vit model?

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ariG23498) [reply src/transformers/models/vit_mae/modeling_tf_vit_mae.py:252]: parity in the API -- the `shape_list` is build explicilty for this operation, it is better if we use it in all the palces.

WDYT?

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ariG23498) [reply src/transformers/models/vit_mae/modeling_tf_vit_mae.py:462]: but this makes no sense -- I strongly believe us raising this concern

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/transformers/models/vit_mae/modeling_tf_vit_mae.py:252]: If that was the case, it would have been used in the rest of the TF models, Isn't it? Or have I missed it?

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/transformers/models/vit_mae/modeling_tf_vit_mae.py:462]: Go ahead.

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/transformers/models/vit_mae/modeling_tf_vit_mae.py:348]: Go ahead.

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ariG23498) [reply src/transformers/models/vit_mae/modeling_tf_vit_mae.py:252]: https://github.com/ariG23498/transformers/blob/915b68287c1a2210aefbcae398f92c139e9a1a4f/src/transformers/models/vit_mae/modeling_tf_vit_mae.py#L408

MY idea about the api comes from this line.

[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/transformers/models/vit_mae/modeling_tf_vit_mae.py:252]: Fair point. But this nowhere maintained i.e., you will see a mix of both `shape_list()` and `.shape`. I will freeze the developments from my end after a while temporarily. I will let you take care of this.

**Comment by ariG23498** [comment]
Hey @sayakpaul do you think we should be changing the pytorch code in the style? My intuition is to not touch it at all.

**Comment by sayakpaul** [comment]
It was a mistake on my end. I will fix it.

**Comment by sayakpaul** [comment]
Done @ariG23498

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py [hunk 12]
```diff
@@ -0,0 +1,1165 @@
...
+        mask_tokens = tf.repeat(
+            self.mask_token,
+            (x.shape[0], ids_restore.shape[1] + 1 - x.shape[1], 1),
+        )
+        x_ = tf.concat([x[:, 1:, :], mask_tokens], acis=1)  # no cls token
+        x_ = tf.gather(
+            x_,
+            axis=1,
+            indices=tf.repeat(ids_restore[..., None], (1, 1, x.shape[2])),
+        )  # unshuffle
```
Comment (ariG23498) [inline src/transformers/models/vit_mae/modeling_tf_vit_mae.py:976]: Should we also try porting this code snippet with the `batch_dims=1`?

Comment (sayakpaul) [reply src/transformers/models/vit_mae/modeling_tf_vit_mae.py:976]: I will test and apply wherever necessary. Hence I mentioned "partially" in the commit message.

**Comment by sayakpaul** [comment]
@ariG23498 with the latest commit (https://github.com/ariG23498/transformers/pull/1/commits/0d993edf1ad5216ec190a614201106801e7d2e16) we should be able to initialize the mode successfully. But the naming needs to be fixed so that cross-loading is successful.

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py [hunk 13]
```diff
@@ -382,7 +382,10 @@ def __init__(self, config: ViTMAEConfig, **kwargs):
             kernel_initializer=get_initializer(config.initializer_range),
             name="value",
         )
-        self.dropout = tf.keras.layers.Dropout(rate=config.attention_probs_dropout_prob)
+        self.dropout = tf.keras.layers.Dropout(
+            rate=config.attention_probs_dropout_prob,
+            name="dropout"
+        )
```
Comment (sayakpaul) [inline src/transformers/models/vit_mae/modeling_tf_vit_mae.py:388]: These should not be formatted as new lines.

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py [hunk 14]
```diff
@@ -455,7 +458,9 @@ def __init__(self, config: ViTMAEConfig, **kwargs):
             kernel_initializer=get_initializer(config.initializer_range),
             name="dense",
         )
-        self.dropout = tf.keras.layers.Dropout(rate=config.hidden_dropout_prob)
+        self.dropout = tf.keras.layers.Dropout(
+            rate=config.hidden_dropout_prob,
+            name="dropout")
```
Comment (sayakpaul) [inline src/transformers/models/vit_mae/modeling_tf_vit_mae.py:463]: Same and elsewhere where applicable.

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py [hunk 15]
```diff
@@ -911,15 +918,15 @@ def call(
...
+        self.decoder_embed = tf.keras.layers.Dense(config.decoder_hidden_size, name="decoder_embed")
 
         decoder_config = deepcopy(config)
         decoder_config.hidden_size = config.decoder_hidden_size
         decoder_config.num_hidden_layers = config.decoder_num_hidden_layers
         decoder_config.num_attention_heads = config.decoder_num_attention_heads
         decoder_config.intermediate_size = config.decoder_intermediate_size
         self.decoder_layers = [
-            TFViTMAELayer(decoder_config, name=f"layers.{j}") for j in range(config.decoder_num_hidden_layers)
+            TFViTMAELayer(decoder_config, name=f"decoder_layers.{j}") for j in range(config.decoder_num_hidden_layers)
```
Comment (sayakpaul) [inline src/transformers/models/vit_mae/modeling_tf_vit_mae.py:929]: Nice work!

File: src/transformers/models/vit_mae/modeling_tf_vit_mae.py (hunk 13 above)
Comment (sayakpaul) [reply src/transformers/models/vit_mae/modeling_tf_vit_mae.py:388]: Maybe run `make style` once?

**Comment by ariG23498** [comment]
@sayakpaul I did not use the formatting yet. Right at the end -- will use the formatter and properly define black styles everywhere -- is it okay with you?

**Comment by sayakpaul** [comment]
Before committing, I find it a good practice to just do `make style` and add the relevant file(s) after that and proceed to commit it. It removes unnecessary file changes from the PR.
